VS 2015 support

rvagg opened the pull request:
Not quite working yet, it gets to the openssl-cli compile and fails with a linking error:

```
apps.obj : error LNK2001: unresolved external symbol ___iob_func [C:\io.js\deps\openssl\openssl-cli.vcxproj]
C:\io.js\Release\openssl-cli.exe : fatal error LNK1120: 1 unresolved externals [C:\io.js\deps\openssl\openssl-cli.vcxproj]
```

This PR upgrades GYP to the latest which understands 2015, and has a lot more changes because there hasn't been an upgrade in a while I suppose.

I've also edited vcbuild.bat to do the appropriate checks.

rvagg commented:
https://jenkins-iojs.nodesource.com/job/iojs+any-pr+multi/85/

which includes a VS2015 host

rvagg commented:
new build: https://jenkins-iojs.nodesource.com/job/iojs+any-pr+multi/88/

I don't understand the Python stack trace for the OSX failures, anybody care to try and enlighten me on that one?

chrisdickinson commented:
Re: python: something in [here](https://github.com/rvagg/io.js/blob/msvs-2015/tools/gyp/pylib/gyp/xcode_emulation.py#L1424-L1468) is adding a non-string-y item to the `env` dict (likely it's adding `None`), and `re.sub` is exploding when it's handed that value.

bnoordhuis commented:
@rvagg Can you mention the gyp commit (e.g. r1234) it's been upgraded to?  I'll see if I can reproduce the OS X failure on my 10.8 machine.

rvagg commented:
@bnoordhuis I would have put a version in there if I knew where to find one, I went looking but alas.. Any hints on where to find a version for this?

bnoordhuis commented:
@rvagg I don't know where you got it from but if it's from subversion, then it's the svn commit.  If it's git, it's the number from git-svn-id.  If this upgrade is to today's HEAD, that would make it r2024.

@chrisdickinson The stack trace suggests it's from [here](https://github.com/rvagg/io.js/blob/msvs-2015/tools/gyp/pylib/gyp/xcode_emulation.py#L1470-L1480) but it looks to me that additional_settings[k] has been sanitized at that point.  Strange.

chrisdickinson commented:
@bnoordhuis I thought so too at first, but [additional_settings gets updated with the contents of env](https://github.com/rvagg/io.js/blob/msvs-2015/tools/gyp/pylib/gyp/xcode_emulation.py#L1477).

rvagg commented:
I printed them out and it is failing on a `None`, `SDKROOT` is the one it's borking on.

rvagg commented:
note this machine only has the commandline tools, not full xcode, which I'm guessing is partly to blame here, still a bug though

rvagg commented:
figured it out, needs a new entry in common.gypi:

```
'SDKROOT': 'macosx'
```

building here: https://jenkins-iojs.nodesource.com/job/iojs+any-pr+multi/90/

bnoordhuis commented:
LGTM.  Maybe you can move the change to common.gypi into a separate commit with an explanation of why it's needed (and move it before the gyp upgrade to keep `git bisect` happy.)

The iojs-win2012r2-msvs2015 failure seems to be related to [this](https://github.com/iojs/io.js/blob/ea7750b/deps/openssl/openssl/e_os.h#L309-333).  I'm not sure why OpenSSL does that but I think we can just rip it out, the only part of OpenSSL that uses stdio (in our configuration) are the apps.

@piscisaureus or @seishun Agree/disagree?

rvagg commented:
Good guess @bnoordhuis, when I fix the define to exclude MSVS 2015 it compiles fine and tests pass with only 1 failure (a TLS failure mind you...). See 7b66043.

Re-running tests in CI here: https://jenkins-iojs.nodesource.com/job/iojs+any-pr+multi/94/nodes=iojs-win2012r2-msvs2015/console

rvagg commented:
oh, that was an all-pass that time, neat.

bnoordhuis commented:
@rvagg It looks like it's running the message tests only.

rvagg commented:
```

rvagg commented:
Ugh, github on phone, I shouldn't try and be productive. Will look in to this tomorrow.

seishun commented:
I don't like the idea of floating patches to OpenSSL to support a preview version of Visual Studio. Maybe it's better to wait for OpenSSL to fix it on their end?

rvagg commented:
Yes, point taken. I was surprised that it wasn't already fixed tbh. I'll leave this PR open so ppl know we are tracking towards it but waiting for it to be easier.

XForce commented:
It is fixed in OpenSSL 1.0.2. Patch is from 2 July 2014 http://git.openssl.org/gitweb/?p=openssl.git;a=commitdiff;h=aab3560b65b9254d17770bb6fe3ca7edd7451429

Fishrock123 commented:
Current OpenSSL 1.0.2 issue: #589

jbergstroem commented:
@rvagg how about rebasing onto the newly landed openssl 1.0.2 v1.x branch and see what jenkins says?

rvagg commented:
good idea, pushed a minimal version that only touches vcbuild.bat, might try it on the 2015 box we have

Fishrock123 commented:
Seems reasonably happy? https://jenkins-iojs.nodesource.com/view/iojs/job/iojs+any-pr+multi/519/

Only thing that stands out is this on win2008r2, which we haven't seen for a little while now:

```
=== release test-tls-ticket-cluster ===
Path: parallel/test-tls-ticket-cluster
[master] got "listening"
(...)
[worker] exit

assert.js:88
  throw new assert.AssertionError({
        ^
AssertionError: Resource leak detected.
    at removeWorker (cluster.js:329:9)
    at ChildProcess.<anonymous> (cluster.js:349:34)
    at ChildProcess.g (events.js:257:16)
    at emitTwo (events.js:87:13)
    at ChildProcess.emit (events.js:169:7)
    at Process.ChildProcess._handle.onexit (child_process.js:1009:12)
Command: c:\workspace\iojs+any-pr+multi\nodes\win2008r2\Release\iojs.exe c:\workspace\iojs+any-pr+multi\nodes\win2008r2\test\parallel\test-tls-ticket-cluster.js
```

rvagg commented:
needs a run on an actual 2015 install first, this only says that it's _probably_ ok with the 2013

Fishrock123 commented:
> needs a run on an actual 2015 install first, this only says that it's probably ok with the 2013

Oh, right. Dx

jasonwilliams200OK commented:
FWIW, libuv v1.6.0+ builds with VS2015 RC.

rvagg commented:
Aye, and openssl 1.0.2 should solve our outstanding problems. Someone should get back on this ...

rvagg commented:
ping @nodejs/build @nodejs/platform-windows, I've just rebased against master and this is working properly now, see https://jenkins-iojs.nodesource.com/job/iojs+any-pr+multi/829/nodes=win2012r2-msvs2015/console for output from a build on the 2015 machine we have set up - I updated to the latest preview of 2015. Any objections against landing this on master?

rvagg commented:
https://github.com/rvagg/io.js/commit/0e54adda1f4a4c75a3bcf9582e7b7dfad2cf5918 is the commit to review, I'll open a new PR against master if there are no objections.

orangemocha commented:
https://github.com/rvagg/io.js/commit/0e54adda1f4a4c75a3bcf9582e7b7dfad2cf5918 LGTM.

You might also want to cherry-pick https://github.com/joyent/node/commit/a58b17456a7fa4ff846c4883b96d6dee7e46a62f. The commit message should be pretty self-explanatory :)

targos commented:
https://github.com/rvagg/io.js/commit/0e54adda1f4a4c75a3bcf9582e7b7dfad2cf5918 that works for me (VS 2015 RC on latest Windows 10 build).

But v8 cannot compile because of a [bug in VS](https://connect.microsoft.com/VisualStudio/feedback/details/1351894). There is a [workaround](https://github.com/v8/v8-git-mirror/blob/master/src/deoptimizer.h#L794-L799) since v4.4.42.

jasonwilliams200OK commented:
@targos, I would not consider that as a bug since in the linked Connect bug report, both OP (David) and VC team member (Jonathan) are in agreement that there is no violation of standard.

Note that the other bug in RC is fixed in recent compiler revision (build 19.00.23008, June 10, 2015) http://webcompiler.cloudapp.net/; it compiles with this code:

```c++
struct User {
  void *operator new(size_t, unsigned);
  void operator delete(void *);
};

int main()
{
    auto q = new (0) User;
    return 0;
}
```

targos commented:
@jasonwilliams200OK I was indeed talking about that other bug. They say it's fixed for RTM but is there a way to get the latest version of the C++ compiler before the release?

jasonwilliams200OK commented:
I think no (that web-compiler is emerged as part of a research project http://rise4fun.com/), but if I venture an educated guess, VS RTM will be available few weeks before Windows 10 release (July 29, 2015), which is about any day now. :)
They are probably also considering to make cl (+link+libs) releases independent of VS release lifecycle, which means we could have win32 (cli-only) server-core running in Docker with cl installed for build stuff.

rvagg commented:
pulled in https://github.com/joyent/node/commit/a58b17456a7fa4ff846c4883b96d6dee7e46a62f as 396b4c6

CI run https://jenkins-iojs.nodesource.com/job/libuv+any-pr+multi/111/

rvagg commented:
derp https://jenkins-iojs.nodesource.com/job/iojs+any-pr+multi/51/

orangemocha commented:
Did something get messed up with this PR? GitHub shows 315 commits and 6686 files changed.

rvagg commented:
yeah, it's really old and targetting the v1.x branch, hence the superfluous commits. It's just the head one that matters.

targos commented:
that's because the PR targets nodejs:v1.x and it is not possible to change the target branch afterwards. Only the last 2 commits are relevant.

srl295 commented:
@rvagg `Intl` won't work under VS2015 because of [the same breakage](http://bugs.icu-project.org/trac/ticket/11798#comment:3) (ICU bug).
